src/messenger.rs

```diff
@@ -549,6 +557,63 @@ where
             }
             return Err(SaslError::ApiError(err));
         }
+
+        Ok(resp)
+    }
+
+    async fn sasl_handshake(&self, mechanism: &str) -> Result<SaslHandshakeResponse, SaslError> {
+        let req = SaslHandshakeRequest::new(mechanism);
+        let resp = self.request(req).await?;
+        if let Some(err) = resp.error_code {
+            return Err(SaslError::ApiError(err));
+        }
+        Ok(resp)
+    }
+
+    pub async fn do_sasl(&self, config: SaslConfig) -> Result<(), SaslError> {
+        let mechanism = config.mechanism();
+        let resp = self.sasl_handshake(mechanism).await?;
+
+        let Credentials { username, password } = config.credentials();
+        let config = SASLConfig::with_credentials(None, username, password).unwrap();
+        let sasl = rsasl::prelude::SASLClient::new(config);
+        let raw_mechanisms = resp.mechanisms.0.unwrap_or_default();
+        let mechanisms = raw_mechanisms
+            .iter()
+            .map(|mech| Mechname::parse(mech.0.as_bytes()).map_err(SaslError::InvalidSaslMechanism))
+            .collect::<Result<Vec<_>, SaslError>>()?;
+        debug!("Supported mechanisms {:?}", mechanisms);
+        let prefer_mechanism =
+            Mechname::parse(mechanism.as_bytes()).map_err(SaslError::InvalidSaslMechanism)?;
+        if !mechanisms.contains(&prefer_mechanism) {
+            return Err(SaslError::UnsupportedSaslMechanism);
+        }
+        let mut session = sasl
+            .start_suggested(&[prefer_mechanism])
+            .map_err(|_| SaslError::UnsupportedSaslMechanism)?;
+        // let selected_mechanism = session.get_mechname();
+        debug!("Using {:?} for the SASL Mechanism", mechanism);
+        let mut data: Option<Vec<u8>> = None;
+
+        // Stepping the authentication exchange to completion.
+        while {
+            let mut out = Cursor::new(Vec::new());
+            // The each call to step writes the generated auth data into the provided writer.
+            // Normally this data would then have to be sent to the other party, but this goes
+            // beyond the scope of this example.
+            let state = session
+                .step(data.as_deref(), &mut out)
+                .expect("step errored!");
+
+            data = Some(out.into_inner());
+
+            // Returns `true` if step needs to be called again with another batch of data.
+            state.is_running()
+        } {
+            let authentication_response = self.sasl_authentication(data.take().unwrap()).await?;
+            data = Some(authentication_response.auth_bytes.0);
+        }
```

Comment:
```suggestion
        // we step through the auth process, starting on our side with NO data received so far
        let mut data_received: Option<Vec<u8>> = None;

        loop {
            let mut to_sent = Cursor::new(Vec::new());
            let state = session
                .step(data_received.as_deref(), &mut to_sent)?;

            if !state.is_running() {
                break;
            }
            
            let authentication_response = self.sasl_authentication(to_sent.into_inner().await?;
            data_received = Some(authentication_response.auth_bytes.0);
        }
```

Some improvements:

- use slightly more readable loop construct
- separate variables for read and write direction (which also eliminates an unwrap)
- replace `except` with proper error handling (you probably need to extend `SaslError`)
- remove the code comment that mentions that you've copied this code from some example :wink:

Reply:
Resolved in [01f1107](https://github.com/influxdata/rskafka/pull/247/commits/01f1107834d6fb3f116e027ae67625e7eb1a9282)